### PR TITLE
Style product

### PR DESCRIPTION
**Pull request by mikeyrayvon**
style single product and implement fixed product header

### CODE REVIEW

File: assets/main.js.liquid [hunk 1]
```diff
@@ -81,33 +81,189 @@ Site.Header = {
...
 
 Site.Product = {
+  minWidth: 720, // minWidth for fixed header layout
+  numberRelatedToPick: 4, // number of related products to show
   init: function() {
     var _this = this;
 
-    if ($('#product-gallery').length) {
-      _this.setGalleryWidth();
+    _this.$container = $('.container'); // used by toggleFixed() on scroll
```
Comment (varffs) [inline assets/main.js.liquid]: This matches 7 items on the page. Not sure if should only be focusing on the main product container?

File: assets/main.js.liquid [hunk 2]
```diff
@@ -81,33 +81,189 @@ Site.Header = {
...
+    // imagesLoaded hasn't finished yet
+    _this.loadedImages = false;
+
+    if (_this.$productGallery.length) {
+      // We have product images, so we do the fancy layout
+      _this.handleLayout();
     }
 
     if ($('#related-products').length) {
+      // Pick a few Related Products to display
```
Comment (varffs) [inline assets/main.js.liquid:104]: when would this not return true?

File: assets/main.js.liquid [hunk 3]
```diff
@@ -81,33 +81,189 @@ Site.Header = {
...
+        // bind imagesLoaded
+        _this.$productGallery.imagesLoaded( function() {
+          // remove min-height for product gallery holder
+          _this.$productGallery.removeClass('min-height');
+          // size image holder to fit image size
+          _this.sizeImageHolder();
+          // get dimensions for toggleFixed()
+          _this.getFixedHeaderDimensions();
+          // toggle fixed header
+          _this.toggleFixed();
```
Comment (varffs) [inline assets/main.js.liquid:132]: have you considered using callbacks for things like this `getFixedHeaderDimensions` and then `toggleFixed`. might make relation between functions clearer to read too

File: assets/main.js.liquid [hunk 4]
```diff
@@ -81,33 +81,189 @@ Site.Header = {
...
+          // init product image zooming
+          _this.bindZoom();
+          // images are loaded
+          _this.loadedImages = true;
+        });
+      } else {
+        // on window resize
+        // if images are already loaded
+        // do the layout again
+        _this.sizeImageHolder();
```
Comment (varffs) [inline assets/main.js.liquid:145]: I think this lot should be bound into a `onResize` type function

File: assets/main.js.liquid [hunk 5]
```diff
@@ -81,33 +81,189 @@ Site.Header = {
...
+    _this.galleryOffset = _this.$productGallery.offset().top;
+    _this.headerHeight = _this.$productHeader.outerHeight(true);
+  },
+
+  toggleFixed: function() {
+    var _this = this;
+
+    var windowScroll = $(window).scrollTop();
+
+    var containerWidth = _this.$container.width();
```
Comment (varffs) [inline assets/main.js.liquid]: can this be checked and cached on resize?

File: assets/main.js.liquid [hunk 6]
```diff
@@ -81,33 +81,189 @@ Site.Header = {
...
+      _this.$productHeader.css('margin-right', 0);
+    }
   },
 
   pickRelated: function() {
-    $('.related-products-item').pick(4);
+    var _this = this;
 
+    // show some related product items
+    $('.related-products-item').pick(_this.numberRelatedToPick);
```
Comment (varffs) [inline assets/main.js.liquid:209]: Thats actually a pretty sweet jQ plugin

File: assets/main.js.liquid [hunk 7]
```diff
@@ -81,33 +81,189 @@ Site.Header = {
...
+
+  sizeImageHolder: function() {
+    // size each image holder to fit image
+
+    $('.product-gallery-image-holder').each(function() {
+      // clear holder height and width
+      $(this).height('auto').width('auto');
+
+      // get image dimensions
+      var $imgElem = $(this).find('.product-gallery-image');
```
Comment (varffs) [inline assets/main.js.liquid]: ends with semi colon but next 2 declared with indent and without var

File: assets/main.js.liquid [hunk 8]
```diff
@@ -81,33 +81,189 @@ Site.Header = {
...
+      // Each image in gallery
+
+      // each image holder has a zoom container element
+      var $zoomContainer = $(this).children('.zoom-container');
+
+      var imgWidth = $(this).width();
+      var imgOffset = $(this).offset();
+
+      // how much are we zooming?
+      var zoomMagnitude = 1.5;
```
Comment (varffs) [inline assets/main.js.liquid]: this could be a setting on the object

**Review by varffs — COMMENTED** [review]
A few code questions/suggestions. 

My only issue is I get layout jump on load. The product info container jumps with the margin on the right hand size doubling in width.

Also it might just be style but the zoom doesnt constrain to the outside of the image. I think I'd prefer how it looks if it did.

File: assets/main.js.liquid (hunk 2 above)
Comment (mikeyrayvon) [reply assets/main.js.liquid:104]: see `snippets/related-products`. there are some conditions for that elem to appear

File: assets/main.js.liquid (hunk 3 above)
Comment (mikeyrayvon) [reply assets/main.js.liquid:132]: I'm not sure how best to do that. you can implement if you want

File: assets/main.js.liquid (hunk 4 above)
Comment (mikeyrayvon) [reply assets/main.js.liquid:145]: within Site.Product?

Comment (mikeyrayvon) [reply assets/main.js.liquid:145]: like Site.Product.onResize()?

File: assets/main.js.liquid (hunk 6 above)
Comment (mikeyrayvon) [reply assets/main.js.liquid:209]: shopify made it cause liquid is so limited

File: assets/main.js.liquid (hunk 7 above)
Comment (mikeyrayvon) [reply assets/main.js.liquid]: whoops

File: assets/main.js.liquid (hunk 4 above)
Comment (varffs) [reply assets/main.js.liquid:145]: That is my personal preference for resize events yes. In this case I'd check against `_this.loadedImages` before doing these functions. My preference for the binding of that actual function is to call `Site.Product.onResize()` inside the `Site.onResize()` (but thats just my way with that not sure if sensible or convoluted)

File: assets/main.js.liquid (hunk 6 above)
Comment (varffs) [reply assets/main.js.liquid:209]: lmao

File: assets/main.js.liquid (hunk 4 above)
Comment (mikeyrayvon) [reply assets/main.js.liquid:145]: i'd just have to check again for the window size and images loaded conditions in the onResize

Comment (varffs) [reply assets/main.js.liquid:145]: yh fair enough. fine as it is

**Comment by mikeyrayvon** [comment]
wait dont merge

**Comment by mikeyrayvon** [comment]
ok go ahead

**Comment by mikeyrayvon** [comment]
ok i'm just showing the product header after the layout is done

**Comment by mixcas** [comment]
👍

**Comment by mixcas** [comment]
i approved this merge

**Pull request merged**